src/Testcontainers/Clients/DockerContainerOperations.cs

```diff
@@ -208,11 +211,41 @@ public async Task<string> RunAsync(IContainerConfiguration configuration, Cancel
         }
       }
 
+      if (configuration.Reuse.HasValue && configuration.Reuse.Value)
+      {
+        var hash = CreateReuseHash(createParameters);
+        createParameters.Labels.Add(TestcontainersClient.TestcontainersReuseHashLabel, hash);
+
+        var containers = await Docker.Containers.ListContainersAsync(new ContainersListParameters()
+        {
+          Filters = new FilterByProperty {
+              { "label", $"{TestcontainersClient.TestcontainersReuseHashLabel}={hash}" }
+            },
+          Limit = 1,
+        }, ct)
+          .ConfigureAwait(false);
+
+        var container = containers.FirstOrDefault();
+        if (container != default) {
+          return container.ID;
+        }
+      }
+
       var createContainerResponse = await Docker.Containers.CreateContainerAsync(createParameters, ct)
         .ConfigureAwait(false);
 
       _logger.DockerContainerCreated(createContainerResponse.ID);
       return createContainerResponse.ID;
     }
+
+    static string CreateReuseHash(CreateContainerParameters createContainerParameters)
+    {
+      var createContainerParametersJson = System.Text.Json.JsonSerializer.Serialize(createContainerParameters);
+
+      using (var sha1 = SHA1.Create())
+      {
+        return Convert.ToBase64String(sha1.ComputeHash(Encoding.Default.GetBytes(createContainerParametersJson)));
+      }
```

Comment:
Is there any preference for the hash algorithm and string format? Wanted to keep it simple, that's why `Convert.ToBase64String` is used here.

Reply:
I have no preference. Maybe we can switch to SHA256, but it probably does not really matter.